gninasrc/lib/cnn_scorer.cpp

```diff
@@ -295,6 +301,110 @@ void CNNScorer::get_net_output(Dtype& score, Dtype& aff, Dtype& loss) {
 
   loss = lossblob->cpu_data()[0];
 }
+
+// Extract ligand atoms and coordinates
+void CNNScorer::setLigand(const model& m){
+
+  if(ligand_atoms.size() == 0){ // Do once at setup
+
+    ligand_atoms.assign(
+      m.get_movable_atoms().cbegin() + m.ligands[0].node.begin,
+      m.get_movable_atoms().cbegin() + m.m_num_movable_atoms
+    );
+  }
+
+  // Coordinates need to be updated at every call
+  ligand_coords.assign(
+      m.coordinates().cbegin() + m.ligands[0].node.begin,
+      m.coordinates().cbegin() + m.m_num_movable_atoms
+  );
+
+  if(!cnnopts.flexopt){ // No flexible residues
+      CHECK_EQ(ligand_atoms.size(), m.get_movable_atoms().size());
+      CHECK_EQ(ligand_coords.size(), m.coordinates().size());
+  }
+
+  CHECK_EQ(ligand_atoms.size(), ligand_coords.size());
+}
+
+// Extracts receptor atoms and coordinates
+// Flex and inflex coordinates are taken from the model's movable atoms
+// Flex coordinates are stored at the beginning, then inflex, then fixed
+void CNNScorer::setReceptor(const model& m){
+
+  if(receptor_atoms.size() == 0){ // Do once at setup
+
+    sz size = m.get_movable_atoms().size() + m.get_fixed_atoms().size();
+
+    receptor_atoms.reserve(size);
+
+    // Insert flexible residues movable atoms
+    receptor_atoms.assign(
+      m.get_movable_atoms().cbegin(), 
+      m.get_movable_atoms().cbegin() + m.ligands[0].node.begin
+    );
+
+    // Get number of movable atoms in flexible residues
+    num_flex_atoms = receptor_atoms.size();
+
+    // Insert inflex atoms
+    receptor_atoms.insert(
+      receptor_atoms.end(), 
+      m.get_movable_atoms().cbegin() + m.m_num_movable_atoms, 
+      m.get_movable_atoms().cend()
+    );
+
+    if(!cnnopts.flexopt){
+      // No atoms from flexible residues should be present
+      CHECK_EQ(receptor_atoms.size(), 0);
+    }
+
+    receptor_atoms.insert( // Insert fixed receptor atoms
+      receptor_atoms.end(),
+      m.get_fixed_atoms().cbegin(),
+      m.get_fixed_atoms().cend()
+    );
+  }
+
+  if(receptor_coords.size() == 0 ){ // Do once at setup
+    sz n = m.get_fixed_atoms().size();
+
+    receptor_coords.reserve(n); // Reserve memory, but size() == 0
+    for(size_t i = 0 ; i < n; i++){
+      const atom& a = m.get_fixed_atoms()[i];
+      const vec& coord = a.coords;
+      receptor_coords.push_back(coord);
+    }
+
+    // Insert inflex
+    receptor_coords.insert(
```

Comment:
inserting at the beginning is less efficient than appending to back of a vector

Reply:
Yes, I intend to change that (see my comment above).